src/main/php/web/auth/SessionBased.class.php

```diff
@@ -86,13 +86,19 @@ public function filter($req, $res, $invocation) {
 
     if (null === $user) {
 
-      // Authentication may require redirection in order to fulfill its job.
-      // In this case, return early from this method w/o passing control on.
-      if (null === ($result= $this->flow->authenticate($req, $res, $session))) return;
-
-      // Otherwise, authorize and transmit session
-      $user= $this->authorize($session, $result);
-      $session->transmit($res);
+      // Authentication may require redirection (indicated by returning NULL)
+      // or showing a page (indicated by returing a Continuation instance) in
+      // order to fulfill its job.
+      $result= $this->flow->authenticate($req, $res, $session);
+      if (null === $result) {
+        return;
```

Comment:
Instead of redirection implemented in the `Flow` class, this could be also refactored to `if ($result instanceof Redirect) { ... }` and then handling that here. This would further simplify the implementations IMO.